packages/beeq/src/components/date-picker/bq-date-picker.tsx

```diff
@@ -0,0 +1,500 @@
+
+    this.bqClear.emit(this.el);
+    this.bqInput.emit({ value: this.value, el: this.el });
+    this.bqChange.emit({ value: this.value, el: this.el });
+    this.inputElem.focus();
+
+    ev.stopPropagation();
+  };
+
+  private handleLabelSlotChange = () => {
+    this.hasLabel = hasSlotContent(this.labelElem);
+  };
+
+  private handlePrefixSlotChange = () => {
+    this.hasPrefix = hasSlotContent(this.prefixElem);
+  };
+
+  private handleSuffixSlotChange = () => {
+    this.hasSuffix = hasSlotContent(this.suffixElem);
+  };
+
+  /**
+   * Generates an array of JSX elements representing calendar months.
+   * If the 'range' property is true and the 'months' property is defined,
+   * it generates multiple calendar months with incremental 'offset' values.
+   * If the 'range' property is false or the 'months' property is undefined,
+   * it generates a single calendar month without an 'offset'.
+   *
+   * @returns An array of JSX elements representing calendar months.
+   */
+  private generateCalendarMonths(): JSX.Element[] {
+    const months: JSX.Element[] = [];
+
+    const commonExportParts = 'heading,table,tr,head,week,th,td';
+    const buttonExportParts = 'button,day,selected,today,disallowed,outside,range-start,range-end,range-inner';
+
+    if (this.range || (this.multi && this.months)) {
+      for (let i = 0; i < this.months; i++) {
+        const offset = i > 0 ? i : undefined;
+        const className = offset ? 'hidden sm:block' : '';
+        months.push(
+          <calendar-month
+            offset={offset}
+            className={className}
+            exportparts={`${commonExportParts},${buttonExportParts}`}
+          />,
+        );
+      }
+    } else {
+      months.push(<calendar-month exportparts={`${commonExportParts},${buttonExportParts}`} />);
+    }
+
+    return months;
+  }
+
+  /**
+   * Processes the focused date value to extract the last date portion.
+   *
+   * @param value - The value to be processed, can be a string, number, or string array.
+   * @returns The extracted last date portion of the value.
+   */
+  private processFocusedDateValue = (value: TInputValue) => {
+    const dateLength = 10; // Length of a standard date in the format YYYY-MM-DD
+
+    if (typeof value === 'string') {
+      return value.slice(-dateLength);
+    }
+    if (typeof value === 'number') {
+      return value.toString();
+    }
+    if (Array.isArray(value)) {
+      return value[value.length - 1];
+    }
+    return null;
+  };
+
+  // render() function
+  // Always the last one in the class.
+  // ===================================
+
+  render() {
+    const labelId = `bq-date-picker__label-${this.name || this.fallbackInputId}`;
+
+    const componentTypes = {
+      multi: 'calendar-multi',
+      range: 'calendar-range',
+      default: 'calendar-date',
+    };
+
+    const key = (this.multi && 'multi') || (this.range && 'range') || 'default';
+    const CalendarComponentType = componentTypes[key];
```

Comment:
This is very nice, well thought 👏🏼🙂 

We can also improve it a bit more if we move the lookup object to a getter and return the corresponding type from there. Something like:

```js
private get CalendarType() {
  // Define a lookup object to map properties to component types
  const componentTypes = {
    multi: 'calendar-multi',
    range: 'calendar-range',
    default: 'calendar-date',
  };
  // Find the first property (multi or range) that is truthy
  const type = ['multi', 'range'].find((t) => this[t]);
  // Return the corresponding component type, or the default type if no truthy property was found
  return componentTypes[type] || componentTypes.default;
}

render() {
  ...
  const CallyCalendar = this.CalendarType;

  ...
  <CallyCalendar
    ...
  >
    ...
  </CallyCalendar>
}
```

What do you think?

Reply:
I would add `as const` for `componentTypes` and type `array` so the return is properly typed.